CONTRIBUTING.md

```diff
@@ -73,48 +96,95 @@ Please note that changes to [lib/recog/version.rb](https://github.com/rapid7/rec
 
-## Updating CPEs
+[^back to top](#contributing-to-recog)
+
+## Fingerprints
+
+### Best Practices
+
+* Create a single fingerprint for each product as long as the pattern remains clear and readable. If that is not possible, the pattern should be logically decomposed into additional fingerprints.
+
+* Create regular expressions that allow for flexible version number matching. This ensures greater probability of matching a product. For example, all known public releases of a product report either `major.minor` or `major.minor.build` format version numbers. If the fingerprint strictly matches this version number format, it would fail to match a modified build of the product that reports only a `major` version number format.
+
+[^back to top](#contributing-to-recog)
+
+### Fingerprint Testing
+
+Once a fingerprint has been added, the `example` entries can be tested by executing `bin/recog_verify` against the fingerprint file:
+
+```shell
+bin/recog_verify xml/ssh_banners.xml
+```
+
+Matches can be tested on the command-line in a similar fashion:
+
+```shell
+$ echo 'OpenSSH_6.6p1 Ubuntu-2ubuntu1' | bin/recog_match xml/ssh_banners.xml -
+MATCH: {"matched"=>"OpenSSH running on Ubuntu 14.04", "service.version"=>"6.6p1", "openssh.comment"=>"Ubuntu-2ubuntu1", "service.vendor"=>"OpenBSD", "service.family"=>"OpenSSH", "service.product"=>"OpenSSH", "os.vendor"=>"Ubuntu", "os.device"=>"General", "os.family"=>"Linux", "os.product"=>"Linux", "os.version"=>"14.04", "service.protocol"=>"ssh", "fingerprint_db"=>"ssh.banner", "data"=>"OpenSSH_6.6p1 Ubuntu-2ubuntu1"}
+```
+
+[^back to top](#contributing-to-recog)
+
+### Updating CPEs
 
 There exists some automation to update the CPEs that might be asserted with
 some recog fingerprints.  This should be run periodically to ensure that all
 fingerprints that could have CPEs do, etc.
 
 First, setup a python3 venv:
 
-  ```
+  ```bash
   python3 -m venv venv
   source venv/bin/activate
   pip install -r requirements.txt
   ```
 
 Download the latest CPE 2.3 dictionary:
 
-  ```
-  wget https://nvd.nist.gov/feeds/xml/cpe/dictionary/official-cpe-dictionary_v2.3.xml.gz
-  ````
+```bash
+wget https://nvd.nist.gov/feeds/xml/cpe/dictionary/official-cpe-dictionary_v2.3.xml.gz
+gunzip official-cpe-dictionary_v2.3.xml.gz
+```
 
 Run the CPE automation against every XML file, using GNU `parallel` to speed things up:
 
-  ```
-  ls xml/*.xml | parallel --gnu "./update_cpes.py {} official-cpe-dictionary_v2.3.xml cpe-remap.yaml && xmllint --format --noblanks {} > {}.bak &&  mv {}.bak {} || echo {}" 2> errors.txt
-  ```
+```bash
+ls xml/*.xml | parallel --gnu "./update_cpes.py {} official-cpe-dictionary_v2.3.xml cpe-remap.yaml && xmllint --format --noblanks {} | sed  ':a;N;$ba;s|</fingerprint>\n  <|</fingerprint>\n\n  <|g;P;D' > {}.bak &&  mv {}.bak {} || echo {}" 2> errors.txt
+```
```

Comment:
This and a similar section below it are the commands to update the CPEs and then format the resulting XML documents.  I had to create a section for `macOS` because I couldn't figure out to do use `sed` to do what I needed in a portable way.